### PR TITLE
Add more warnings to for the ObjC runtime build

### PR DESCRIPTION
**Pull request by thomasvl**
Working on https://github.com/google/protobuf/issues/1599, specifically:
- Turn on more warnings that the Xcode UI calls out with individual controls.
- Manually add:
  -Wundef
  -Wswitch-enum
- Manually add and then diable in the unittests because of XCTest's headers:
  -Wreserved-id-macro
  -Wdocumentation-unknown-command
- Manually add -Wdirect-ivar-access, but disable it for the unittests and in
  the library code (via #pragmas to suppress it). This is done so proto users
  can enable the warning.

### CODE REVIEW

**Comment by thomasvl** [comment]
@sergiocampama fyi

File: src/google/protobuf/compiler/objectivec/objectivec_message.h [hunk 1]
```diff
@@ -66,6 +66,9 @@ class MessageGenerator {
   void GenerateExtensionRegistrationSource(io::Printer* printer);
   void DetermineForwardDeclarations(set<string>* fwd_decls);
 
+  // Checks if the message or a nested message includes a oneof definition.
+  bool IncludesOneOfDefinition() const;
```
Comment (sergiocampama) [inline src/google/protobuf/compiler/objectivec/objectivec_message.h]: this doesn't look to be related to this change :p

Comment (thomasvl) [inline src/google/protobuf/compiler/objectivec/objectivec_message.h]: It is, it is used to pick off when to add the extra pragma ignore to the generated sources (direct ivar access)

Comment (sergiocampama) [inline src/google/protobuf/compiler/objectivec/objectivec_message.h]: ahh gotcha...
